javalib/src/main/scala/java/util/ArrayList.scala

```diff
@@ -14,80 +14,187 @@ package java.util
 
 import java.lang.Cloneable
 import java.lang.Utils._
+import java.util.ScalaOps._
 
 import scala.scalajs._
+import scala.scalajs.LinkingInfo.isWebAssembly
 
-class ArrayList[E] private (private[ArrayList] val inner: js.Array[E])
+class ArrayList[E] private (innerInit: AnyRef, private var _size: Int)
     extends AbstractList[E] with RandomAccess with Cloneable with Serializable {
   self =>
 
+  /* This class has two different implementations for handling the
+   * internal data storage, depending on whether we are on Wasm or JS.
+   * On JS, we utilize `js.Array`. On Wasm, for performance reasons,
+   * we avoid JS interop and use a scala.Array.
+   * The `_size` field (unused in JS) keeps track of the effective size
+   * of the underlying Array for the Wasm implementation.
+   */
+
+  private val innerJS =
+    if (isWebAssembly) null
+    else innerInit.asInstanceOf[js.Array[E]]
+
+  private var innerWasm =
+    if (!isWebAssembly) null
+    else innerInit.asInstanceOf[Array[Any]]
+
   def this(initialCapacity: Int) = {
-    this(new js.Array[E])
+    this(
+      if (isWebAssembly)
+        new Array[Any](Math.max(initialCapacity, 0))
+      else new js.Array[E],
```

Comment:
```suggestion
      if (isWebAssembly) new Array[Any](Math.max(initialCapacity, 0))
      else new js.Array[E],
```
To avoid the `Math.max`, though, we can move the `initialCapacity < 0` check before the super call, like this:
```scala
  def this(initialCapacity: Int) = {
    this(
      {
        if (initialCapacity < 0)
          throw new IllegalArgumentException
        if (isWebAssembly) new Array[Any](initialCapacity)
        else new js.Array[E],
      },
      0
  }
```